sns/parity_math.c

```diff
@@ -385,85 +768,911 @@ static void xor_calculate(struct m0_parity_math *math,
+		   m0_parity_elem_t alpha)
+{
+	uint32_t                i;
+	uint32_t                seg_size;
+	uint8_t                *y_addr;
+	uint8_t                *x_addr;
+	m0_bcount_t             step;
+	struct m0_bufvec_cursor x_cursor;
+	struct m0_bufvec_cursor y_cursor;
+
+	M0_ENTRY("y=%p, x=%p, alpha=%u", y, x, (uint32_t)alpha);
+
+	M0_PRE(y != NULL && x != NULL);
+	M0_PRE(y->ov_vec.v_nr != 0);
+	M0_PRE(y->ov_vec.v_nr == x->ov_vec.v_nr);
+	M0_PRE(y->ov_vec.v_count[0] != 0);
+	M0_PRE(y->ov_vec.v_count[0] == x->ov_vec.v_count[0]);
+	seg_size = y->ov_vec.v_count[0];
+	M0_PRE(m0_forall(i, y->ov_vec.v_nr, y->ov_vec.v_count[i] == seg_size));
+
+	m0_bufvec_cursor_init(&y_cursor, y);
+	m0_bufvec_cursor_init(&x_cursor, x);
+	do {
+		x_addr  = m0_bufvec_cursor_addr(&x_cursor);
+		y_addr  = m0_bufvec_cursor_addr(&y_cursor);
+
+		switch (alpha) {
+		/* This is a special case of the 'default' case that follows.
+		 * Here we avoid unnecessary call to gmul.*/
+		case 1:
+			for (i = 0; i <seg_size; ++i) {
+				y_addr[i] = gadd(y_addr[i],
+						 x_addr[i]);
+			}
+			break;
+		default:
+			for (i = 0; i < seg_size; ++i) {
+				y_addr[i] = gadd(y_addr[i], gmul(x_addr[i],
+							         alpha));
+			}
+			break;
+		}
+		step = m0_bufvec_cursor_step(&y_cursor);
+	} while (!m0_bufvec_cursor_move(&x_cursor, step) &&
+		 !m0_bufvec_cursor_move(&y_cursor, step));
+
+	M0_LEAVE();
+}
+
+static bool is_usable(const struct m0_sns_ir *ir,
+		      const struct m0_bitmap *in_bmap,
+		      struct m0_sns_ir_block *failed_block)
+{
+	size_t   i;
+	uint32_t last_usable_bid;
+
+	M0_PRE(in_bmap != NULL && failed_block != NULL && ir != NULL);
+	M0_PRE(in_bmap->b_nr == failed_block->sib_bitmap.b_nr);
+
+	last_usable_bid = last_usable_block_id(ir, failed_block->sib_idx);
+	if (last_usable_bid == ir_blocks_count(ir))
+		return false;
+	for (i = 0; i <= last_usable_bid; ++i) {
+		if (m0_bitmap_get(in_bmap, i) &&
+		    !m0_bitmap_get(&failed_block->sib_bitmap, i))
+			return false;
+	}
+	return true;
+}
+
+static inline bool is_valid_block_idx(const struct m0_sns_ir *ir,
+				      uint32_t block_idx)
+{
+	return block_idx < ir_blocks_count(ir);
+}
+
+#if ISAL_ENCODE_ENABLED
+static void reed_solomon_fini(struct m0_parity_math *math)
+{
+	M0_ENTRY();
+	m0_free(math->pmi_rs.rs_encode_matrix);
+	m0_free(math->pmi_rs.rs_encode_tbls);
+	m0_free(math->pmi_rs.rs_decode_tbls);
+	m0_free(math->pmi_rs.rs_alive_idx);
+	m0_free(math->pmi_rs.rs_failed_idx);
+	m0_free(math->pmi_rs.rs_bufs_in);
+	m0_free(math->pmi_rs.rs_bufs_out);
+	M0_LEAVE();
+}
+
+static int reed_solomon_init(struct m0_parity_math *math)
+{
+	struct m0_reed_solomon *rs;
+	uint32_t                total_count;
+	uint32_t                tbl_len;
+	int                     ret = 0;
+
+	M0_ENTRY("math=%p", math);
+
+	M0_PRE(parity_math_invariant(math));
+
+	rs = &math->pmi_rs;
+	total_count = math->pmi_data_count + math->pmi_parity_count;
+	tbl_len = math->pmi_data_count * math->pmi_parity_count * 32;
```

Comment:
What is the significance of 32 here, can we define a macro for it ?

Reply:
Addressed. Added relevant comment as well.